index.js

```diff
@@ -20,6 +22,110 @@ app.get('/info', (req, res) => {
   res.send({ serverName: packageData.name, serverVersion: packageData.version });
 });
 
+// route for displaying all products
+app.get('/products/all', (req, res) => {
+
+  var jsonOutputObject = {
+    products : []
+  };
+
+  for(var i = 0; i < productData.products.length; i++) {
+    
+    jsonOutputObject.products.push(productData.products[i])
+
+    // Getting category name for the product
+    for(var j = 0; j < categoryData.categories.length; j++) {
+      if(jsonOutputObject.products[i].categoryId == categoryData.categories[j].id) {
+        jsonOutputObject.products[i].categoryName = categoryData.categories[j].categoryName;
+        break;
+      }
+    }
+  }
+
+  res.send(jsonOutputObject);
+});
+
+// Getting product info from its ID
+app.get('/product/:id', (req, res) => {
+
+  let userInput = req.params.id;
+  var productId;
+  var jsonOutputObject = null;
+
+  // Generating valid ID and returning error message when ID is invalid
+  if (userInput < 1) {
+    res.send('Invalid product ID');
+    return;
+  } else {
+    productId = `P${userInput.padStart(3, '0')}`
+  }
+
+  for(var i=0; i<productData.products.length; i++) {
+    if(productData.products[i].id == productId) {
+      jsonOutputObject = productData.products[i];
+      break;
+    }
+  } 
+
+  // Checking for null at the end would cause NullPointerException when trying to get the categoryName
+  if(jsonOutputObject == null) {
+    res.send('Product not found!');
+    return;
+  }
+
+  // Getting category name
+  for(var i = 0; i < categoryData.categories.length; i++) {
```

Comment:
Use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of